Zend/zend_execute.c

```diff
@@ -1010,8 +1010,12 @@ static zend_never_inline zval* zend_assign_to_typed_prop(zend_property_info *inf
 	zval tmp;
 
 	if (UNEXPECTED(info->flags & ZEND_ACC_READONLY)) {
-		zend_readonly_property_modification_error(info);
-		return &EG(uninitialized_zval);
+		if (Z_PROP_FLAG_P(property_val) & IS_PROP_REINITABLE) {
+			Z_PROP_FLAG_P(property_val) &= ~IS_PROP_REINITABLE;
```

Comment:
This will remove the flag before type checks (`i_zend_verify_property_type`). This means the property will become unassignable thereafter. Is that consistent with how this behaves otherwise?

Reply:
Hm, yes, the goal is to lock any modification after a maximum of one assignment during `__clone()`. Thereafter readonly properties mustn't be modified until they are not unlocked again.

Reply:
Hm... Sorry, half of my comment is missing. What I meant is that this will lock the property even if there was no assignment, i.e. when `i_zend_verify_property_type` fails.

Reply:
Ah, I wasn't able to read between the lines, but you are completely right! a failed assignment attempt should be able to be retried.

Reply:
Hmm, but then the original code would also be buggy, since readonly properties couldn't be assigned to after a type error. I haven't yet got the time to debug what kind of other code runs instead of this one, but according to the new test I added, this should be fine as is.

Reply:
> Hmm, but then the original code would also be buggy, since readonly properties couldn't be assigned to after a type error.

Yeah, that's what I tried to say. Look at this incorrect test here.

```php
--TEST--
Test that readonly property isn't locked after unsuccessful assignment
--FILE--
<?php

class Foo {
    public function __construct(
        public readonly int $bar,
    ) {}

    public function __clone()
    {
        try {
            $this->bar = [];
        } catch (Throwable) {}

        $this->bar = 2;
    }
}

$foo = new Foo(1);
var_dump(clone $foo);

?>
--EXPECTF--
Fatal error: Uncaught Error: Cannot modify readonly property Foo::$bar in %s:%d
Stack trace:
#0 %s(%d): Foo->__clone()
#1 {main}
  thrown in %s on line %d
```

Reply:
I think I managed to fix the problem, although it required the same change in `zend_std_write_property`